library/ssl_tls13_client.c

```diff
@@ -2120,6 +2120,34 @@ static int ssl_parse_server_psk_identity_ext( mbedtls_ssl_context *ssl,
 
 #endif
 
+/* Early Data Extension
+*
+* struct {} Empty;
+*
+* struct {
+*   select (Handshake.msg_type) {
+*     case new_session_ticket:   uint32 max_early_data_size;
+*     case client_hello:         Empty;
+*     case encrypted_extensions: Empty;
+*   };
+* } EarlyDataIndication;
+*/
+int ssl_parse_early_data_ext(mbedtls_ssl_context *ssl,
+    const unsigned char *buf, size_t len) {
+  ( ( void )buf );
+  if (ssl->state == MBEDTLS_SSL_ENCRYPTED_EXTENSIONS
+      && ssl->handshake->early_data == MBEDTLS_SSL_EARLY_DATA_ON
+      && len == 0) {
+    ssl->early_data_status = MBEDTLS_SSL_EARLY_DATA_ACCEPTED;
+    return 0;
+  }
+  return MBEDTLS_ERR_SSL_INTERNAL_ERROR;
```

Comment:
This is poorly documented, but we aim to use `MBEDTLS_ERR_SSL_INTERNAL_ERROR` like assertion failures -- in particular, they should _never_ trigger in a sane Mbed TLS built, regardless of user input. 

A more suitable error code might be `MBEDTLS_ERR_SSL_BAD_INPUT_DATA` here.